stdlib/asyncio/coroutines.pyi

```diff
@@ -1,7 +1,9 @@
-from typing import Any, Callable, TypeVar
+import types
+from typing import Any, Callable, Coroutine, TypeVar
+from typing_extensions import TypeGuard
 
 _F = TypeVar("_F", bound=Callable[..., Any])
 
 def coroutine(func: _F) -> _F: ...
 def iscoroutinefunction(func: object) -> bool: ...
-def iscoroutine(obj: object) -> bool: ...
+def iscoroutine(obj: object) -> TypeGuard[types.GeneratorType[Any, Any, Any] | Coroutine[Any, Any, Any]]: ...
```

Comment:
As a note, in https://github.com/encode/httpx/pull/2269 where we have this kind of usage:

```python
response = some_func()

if asyncio.iscoroutine(response):
    response = await response
```

We had to narrow down to `TypeGuard[Coroutine]` ourselves because on py3.7 `mypy` would raise an error due to `Generator` not awaitable.

I think I would have agreed with this:

> In this case, considering the deprecation of generator coroutines, using TypeGuard[Coroutine] seems like the better option. Primer output actually proves that using a union return type is not serving our users well.

Reply:
Thanks, raised it here #8114

Reply:
Actually it's here #8104